### DISCUSSION TITLE
Re-indexing does not remove photos from index if originals folder is empty

### DISCUSSION THREAD

**Discussion by schaefsteven** [photoprism/photoprism#3938] [Q&A]
#### 1. What is not working as documented?

If all folders and files are deleted from the originals folder, re-indexing does not remove them from the index. 

#### 2. How can we reproduce it?

1. Have indexed content in your originals folder
2. Using something other than Photoprism (a file manager for example) delete ALL of the content in the originals folder so that it is an empty directory
3. Run either Index or Complete Rescan

#### 3. What behavior do you expect?

The library should now be empty. Instead, all of the content still appears in the library. 

#### 4. What could be the cause of your problem?

Seems like the indexing process doesn't run properly because it doesn't have a single file to work with, so it just decides not to run. I tried the same process but deleting all but one photo, and things work as expected. Also, if I import new content, the old deleted items get removed from the library. So it seems that indexing just doesn't work if the originals folder is empty. 

#### 5. Can you provide us with example files for testing, error logs, or screenshots?

log: 
```
2023-05-23 10:31:36 INFO indexing completed in 0 s

2023-05-23 10:31:36 INFO index: updated 0 files [417.238µs]

2023-05-23 10:31:36 INFO originals folder is empty

2023-05-23 10:31:36 INFO indexing originals...
```
#### 6. Which software versions do you use?

PhotoPrism® CE
Build 230514-0b780defb

Installed on TrueNAS SCALE via TrueCharts
14.0.17

Possibly related issue: #1010

**Comment by graciousgrey** [comment]
This is a security mechanism. We implemented it so that indexing does not remove all files from the index in case for example the originals folder cannot be accessed due to network problems. 

--> https://github.com/photoprism/photoprism/issues/2299

**Comment by schaefsteven** [comment]
> This is a security mechanism. We implemented it so that indexing does not remove all files from the index in case for example the originals folder cannot be accessed due to network problems.
> 
> --> #2299

Got it. I think it would be a good idea then to indicate this behavior in A) the logs, and B) the message at the bottom of the screen when indexing completes. Something like "Index not updated: Originals folder is empty."

**Comment by lastzero** [comment]
You should see something like this in the logs. As most people don't index an empty library and there is a dedicated reset command, PhotoPrism currently does not feature an info message directly on the index page in the UI. Note that all text in the UI must be translated in over 40 languages.

**Comment by lastzero** [comment]
Feel free to send a pull request if "originals folder is empty" isn't explicit enough. It used to be a warning but this caused quite a lot of noise on the Library > Errors page for some users as in the beginning, there often really aren't any files, so it's not a problem.

**Comment by artincube** [comment]
I opened a thread in discussions about something in relation:

I would like to be able to work metadata "offline", that is launch Photoprism without Originals drive mounted and work metadata on the go.

if I unplug the originals drive while Photoprism is running,
I get an empty originals folder in the web ui, but still with the proper number of items in the sidebar.

**Comment by graealex** [comment]
You can just put a single random cat picture in your originals folder and re-index, and then it will purge your cache if you also selected "Cleanup" as the indexing option.

I agree that it is a good safety mechanism to prevent your whole index getting purged just because there was some trouble accessing the originals folder.
